mm2src/common/wasm_indexed_db.rs

```diff
@@ -0,0 +1,1141 @@
+    pub fn new(db_name: &str) -> IndexedDbBuilder {
+        IndexedDbBuilder {
+            db_name: db_name.to_owned(),
+            db_version: 1,
+            tables: HashMap::new(),
+        }
+    }
+
+    pub fn with_version(mut self, db_version: u32) -> IndexedDbBuilder {
+        self.db_version = db_version;
+        self
+    }
+
+    pub fn with_table<Table: TableSignature>(mut self) -> IndexedDbBuilder {
+        let on_upgrade_needed_cb = Box::new(Table::on_upgrade_needed);
+        self.tables.insert(Table::table_name().to_owned(), on_upgrade_needed_cb);
+        self
+    }
+
+    pub async fn init(self) -> InitDbResult<IndexedDb> {
+        Self::check_if_db_is_not_open(&self.db_name)?;
+        let (table_names, on_upgrade_needed_handlers) = Self::tables_into_parts(self.tables)?;
+
+        let window = web_sys::window().expect_w("!window");
+        let indexed_db = match window.indexed_db() {
+            Ok(Some(db)) => db,
+            Ok(None) => return MmError::err(InitDbError::NotSupported("Unknown error".to_owned())),
+            Err(e) => return MmError::err(InitDbError::NotSupported(stringify_js_error(&e))),
+        };
+
+        let db_request = match indexed_db.open_with_u32(&self.db_name, self.db_version) {
+            Ok(r) => r,
+            Err(e) => return MmError::err(InitDbError::InvalidVersion(stringify_js_error(&e))),
+        };
+        let (tx, mut rx) = mpsc::channel(1);
+
+        let onerror_closure = construct_event_closure(DbOpenEvent::Failed, tx.clone());
+        let onsuccess_closure = construct_event_closure(DbOpenEvent::Success, tx.clone());
+        let onupgradeneeded_closure = construct_event_closure(DbOpenEvent::UpgradeNeeded, tx.clone());
+
+        db_request.set_onerror(Some(onerror_closure.as_ref().unchecked_ref()));
+        db_request.set_onsuccess(Some(onsuccess_closure.as_ref().unchecked_ref()));
+        db_request.set_onupgradeneeded(Some(onupgradeneeded_closure.as_ref().unchecked_ref()));
+
+        let mut on_upgrade_needed_handlers = Some(on_upgrade_needed_handlers);
+        while let Some(event) = rx.next().await {
+            match event {
+                DbOpenEvent::Failed(e) => return MmError::err(InitDbError::OpeningError(stringify_js_error(&e))),
+                DbOpenEvent::UpgradeNeeded(event) => {
+                    Self::on_upgrade_needed(event, &db_request, &mut on_upgrade_needed_handlers)?
+                },
+                DbOpenEvent::Success(_) => {
+                    let db = Self::get_db_from_request(&db_request)?;
+                    Self::cache_open_db(self.db_name.clone());
+
+                    return Ok(IndexedDb {
+                        db,
+                        db_name: self.db_name,
+                        tables: table_names,
+                    });
+                },
+            }
+        }
+        unreachable!("The event channel must not be closed before either 'DbOpenEvent::Success' or 'DbOpenEvent::Failed' is received");
```

Comment:
I suspect that in case of graceful MM2 shutdown (e.g. users log out and then log in using a different seed) this can potentially panic. Please also ensure that we won't have panics in other places in case of `stop` RPC is called. Also, if we will have problems with a graceful shutdown, we can advise GUIs to forcefully refresh the page on log out. This comment is not an approval blocker.